h4l/production/invariant_mass.py

```diff
@@ -14,42 +14,52 @@
 
 @producer(
     uses=(
-        # "nJet", "Jet.pt", "Jet.eta", "Jet.phi", "Jet.mass", "Jet.area", 
+        # "nJet", "Jet.pt", "Jet.eta", "Jet.phi", "Jet.mass", "Jet.area",
         # "Jet.rawFactor", "Jet.jetId",
-        
-        {f"{field}.{var}"
-         for field in ["Electron", "Muon"]
-         for var in ["pt", "mass", "eta", "phi", "charge"]
-         }|
-         {
-             attach_coffea_behavior,
-         }
+        {
+            f"{field}.{var}"
+            for field in ["Electron", "Muon"]
+            for var in ["pt", "mass", "eta", "phi", "charge"]
+        } | {
+            attach_coffea_behavior,
+        }
     ),
     produces={
         "m4l",
     },
 )
 def four_lep_invariant_mass(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    """
+    Construct four-lepton invariant mass given the Electron and Muon arrays.
+    """
 
-    # category ids
-    # from IPython import embed
-    # embed()
+    # attach coffea behavior for four-vector arithmetic
     events = self[attach_coffea_behavior](
         events,
         collections=["Electron", "Muon"],
-        **kwargs
+        **kwargs,
     )
-    from IPython import embed; embed()
 
-    diele = events.Electron[:, :2].sum(axis=1)
-    dimuon = events.Muon[:, :2].sum(axis=1)
-    fourlep = diele + dimuon
-    diele_mask = ak.num(events.Electron, axis=1) >= 2
-    dimuon_mask = ak.num(events.Muon, axis=1) >= 2
+    # pad the lepton arrays with `None` entries
+    # to ensure they have at least length 2
+    electron = ak.pad_none(events.Electron, 2, axis=1)
+    muon = ak.pad_none(events.Muon, 2, axis=1)
+
+    # sum over first two elements of each collection
+    # (`None` entries propagate)
+    dielectron = electron[:, :2].sum(axis=1)
+    dimuon = muon[:, :2].sum(axis=1)
+
+    # sum the results to form the four-lepton four-vector
+    fourlep = dielectron + dimuon
+
+    # write out the resulting mass to the `events` array,
+    # replacing `None` values with a predefined EMPTY_FLOAT value
     events = set_ak_column_f32(
         events,
         "m4l",
-        ak.where(diele_mask&dimuon_mask, fourlep.mass, EMPTY_FLOAT),
+        ak.fill_none(fourlep.mass, EMPTY_FLOAT),
```

Comment:
I don't think this is correct. Since you padded the arrays before, you would now also consider the masses of single leptons since the sum operation of `Lepton + None` returns just the lepton again. We only want to consider cases where we actually have four leptons though, right?

Reply:
But `Lepton + None` should return `None`, no? If that's not the case I agree this would need to be changed, but otherwise the m4l should only be filled if there are >=4 leptons. To be tested.

Reply:
OK, I just tested this and turns our you're right, adding a missing value is just a no-op. My bad, I was operating under the assumption that missing values behave like `np.nan`, which always propagates through basic arithmetic operations. I'll add back the lepton number check in that case.